### DISCUSSION TITLE
How to directly inject value from a service method to a twig global variable ?

### DISCUSSION THREAD

**Discussion by thgauthier** [symfony/symfony#45024] [Q&A] [answered]
I know it is possible to inject a service directly to a global variable by setting a service :
![Capture d’écran du 2022-01-14 09-49-40](https://user-images.githubusercontent.com/95478233/149486402-a1d4d9ee-8f03-4d7d-8652-6431356ef808.png)

But in case my getColor() function return directly the color without need to provide any parameters :
![Capture d’écran du 2022-01-14 10-02-30](https://user-images.githubusercontent.com/95478233/149488754-efed9517-187c-4ab0-96b8-c3fa1c5f55b5.png)

Suppose my getColor() function return value from database query, I don't what to query each time the DB by calling 
{{ dump(main_color.getColor()) }}

Any idea ?
Thanks

**Comment by stof** [comment, marked as answer]
I don't think you want to query the database either at the time of instantiating Twig to get the global. This would query the DB even when not using this color, and might query it too early.
If you don't want to query each time it is called, use memoization in your method.

**Comment by toto33331** [comment]
If someone is still annoyed by this problem, here is my solution (for a menu in SF6.4)

config/packages/twig.yaml :
```
twig:
    globals:
        global_menu: '@App\Service\MenuService'
```

src/Service/MenuService.php
```
<?php

namespace App\Service;

use App\Entity\Menu;
use Doctrine\ORM\EntityManagerInterface;

class MenuService
{
    private $manager;
    public function __construct(EntityManagerInterface $manager)
    {
        $this->manager = $manager;
    }
    
    public function getMenu(): array
    {
        $manager = $this->manager;
        $menuRepo = $manager->getRepository(Menu::class);
        $menuItems = $menuRepo->findBy(array(), array('position'=>'asc'));
        return $menuItems;
    }
}
```

templates/layout.html.twig:

```
<ul class="nav">
    {% for item in global_menu.menu %}
      <li class="nav_sub_list_item">
        <a href="{{ item.link }}">{{ item.text }}</a>
      </li>
    {% endfor %}
  </ul>

```

As you see, my example is not optimised since it will get the records from the database on every page. You can easily cache this. I just didn't add the optimised code here to get the simplest code.
